observability (#1449)
tigera/docs: use-cases/observability.mdx

---
description: A guide for using Calico's observability tools.
title: Observability
---

# Observability

This guide will walk you through the different observability and monitoring capabilities in Calico so you can learn how to observe and troubleshoot workload communications, performance, and operations in a Kubernetes cluster.

## Overview

### What is observability?

> Observability is the ability to understand the internals of a system by analyzing the internals of a system. 

Reviewer: ```suggestion
People use observability tools to understand a complex system by visually analyzing what's going on in that system.
```
A suggestion.

In the context of Kubernetes, that system would be the cluster, and the internals could be nodes, pods, resources, network policies, and so on.
In complex systems, with lots of dynamic, interconnected parts, observability puts a visual frontend on what would otherwise likely be a series of recursive commands in a CLI to obtain the same information.
What that front-end is will depend on the software that’s integrated within the cluster, and will likely depend on which internals are being shown, and who the target audience is.
One example might be a dashboard, showing various metrics to an end-user who needs an overview of a system to monitor it’s health. 

> Observability could also go more granular than that, representing visually how different resources installed in the cluster are connected and dependent on each other.

Reviewer: ```suggestion
At a more granular level, observability tools can help you visualize how different resources installed in the cluster are connected and dependent on each other.
```
A suggestion

> In summary, observability is a way to visualize, organize and understand what’s going on in our Kubernetes clusters.

Reviewer: rm

Flow Logs are available through the Dynamic Service and Threat Graph, and automatically filter based on the view or selected resource.

![Service graph flow logs](/img/use-cases/service-graph-flow-logs.png)

The default view is to have the flow metadata collapsed, with the flow logs table configured with different columns to allow you to choose a table summary that best suits your needs.

Logs can also be viewed and queried through Kibana for more advanced use cases, and will be covered in the next section.
Flow metadata provides an extra level of detail for understanding, investigating, and troubleshooting network flow issues, such as:

* Identifying policy impact on traffic flow: which policies are evaluating traffic flow, in which order, is the flow allowed or denied
* Which processes are initiating flows
* Issues where connections could not be established
* Changes in application or service load 
* Performance issues with latency or packets dropping
* Knowing which IP addresses or FQDNs are communicating with your cluster

Specific examples of how to use these observability features for these issues can be found below.

> ### Most-detailed observability

Reviewer: This title reads a bit awkward. I think you should ditch the word 'observability' as you did with the previous titles. 

Perhaps something along these lines?

* Fine-grained log analysis
* Visualizing logs
* Observing detailed log output

##### Inbound and outbound bytes

Visualizing inbound and outbound bytes may help identify whether your application is experiencing higher than normal load or whether there is malicious activity causing increased traffic.

* *Dynamic Service and Threat Graph* - Hovering over flow lines between objects in Dynamic Service and Threat Graph brings up a popup that reports the number of allowed or denied packets and bytes.
* *Flow Viz* - Flow Viz by default is a volumetric view of traffic that helps your identify which namespaces or services are generating the most traffic. The right-hand panel in Flow Viz also displays the CPS, PP, and BPS.
* *Logs* - Flow logs record the number of bytes in and bytes out for each flow.
* *Kibana* - Dashboards in Kibana have pie charts that show volumetric traffic by bytes in and out for source and destination namespaces.
  Dashboards can be filtered to provide more detailed information.

[See how filtering can be set up in dashboards to view bytes in/out for specific namespaces and find flows of interest](https://fast.wistia.com/embed/channel/lhjf79y3oy?wchannelid=lhjf79y3oy&wmediaid=6qmpv3il71).

##### Identify flows with “bytes_in : 0”

> Flows with “bytes_in : 0” and “action : allow” indicate that an upstream firewall denied a flow, or the server is not responding to the connection.

Reviewer: For code snippets, please format with backticks:
```
`bytes_in : 0`
```

#### Flow properties and metadata

##### Identify flows to and from specific processes

Within a pod there could be more than one container, and each container could have different processes initiating or receiving flows.
Typically, identifying the port or container level might be easier, but identifying the specific process is harder.
Calico uses eBPF probes to record this data and enrich flow logs with a process ID. All Calico products (Open Source, Enterprise or Cloud) support eBPF as a dataplane.

This is useful for identifying a workload initiating communication with decommissioned services or services it shouldn’t be communicating with.
It can also be used to identify any communication related to a vulnerability, such as log4j. 

* *Logs* - Flow logs contain process metadata such as process ID, name, and process arguments.
* *Kibana* - Dashboards and logs in Kibana can be queried using kql to search for flows with specific process ids, names or arguments.
  This will return all flows that match, helping to identify the source and destination of flows for those processes.

> To view an example using Kibana to filter and identify flows related to log4j communication, watch [this video](https://fast.wistia.com/embed/channel/lhjf79y3oy?wchannelid=lhjf79y3oy&wmediaid=ooka4wvfxz)

Reviewer: Revisit anchor text.

This makes it easy to identify which ports are being communicated to from within the cluster.
You could, for example, identify all traffic communicating with a database, or identify SSH traffic.
This helps you identify workloads communicating over particular service ports.

[See how to identify traffic to specific ports in Kibana](https://fast.wistia.com/embed/channel/lhjf79y3oy?wchannelid=lhjf79y3oy&wmediaid=0n4rioj2jk)

##### Identify traffic to specific FQDNs

If workloads are communicating with external services where the IP address correlates with a DNS entry, Calico Enterprise and Calico Cloud can record the fully qualified domain name (FQDN) of that service.
This helps you identify workloads that are communicating with public external services.

* *Logs* - Flow logs contain the FQDN.
  The destination will show as "pub", and the dest_domains field will show the domain name based on a DNS lookup.
* *Kibana* - The Kibana dashboards have a "unique-domains" widget that lists the top values of dest_domains from flow logs, with a record count.
  Clicking on a domain name in that widget will filter all flows that sent traffic to that FQDN.

> To see this in Kibana, watch [this video](https://fast.wistia.com/embed/channel/lhjf79y3oy?wchannelid=lhjf79y3oy&wmediaid=huq9luzvux)

Reviewer: Revisit this anchor text
